### ISSUE TITLE
Implement multidimensional discrete distribution

### ISSUE THREAD

**Issue by rcurtin** [mlpack/mlpack#810]
See #809 for the discussion that led to this ticket.

Currently, the `DiscreteDistribution` class in `src/mlpack/core/dists/` only supports single-dimension discrete inputs (like `(1)`, `(2)`, `(5)`, etc.).  It would be very useful if it instead supported multi-dimensional discrete inputs (like `(1, 2, 3)`, `(3, 5, 8)`, `(1, 7, 5)`).  So the `DiscreteDistribution` class can easily be extended to support that case.  Right now, in the single dimension case, the probability of some input X is simply stored as P(X).  Now, with multi-dimensional inputs x_1, x_2, ..., x_d, the probability should be P(x_1) P(x_2) ... P(x_d) (just single-dimension probabilities multiplied).

When we close this ticket it should be possible to train multi-dimensional discrete HMMs with the `mlpack_hmm_train` program, and there should also be tests for the multi-dimensional support in `src/mlpack/tests/distribution_test.cpp`.

**Comment by MikeLing** [comment]
Hey @rcurtin , I would like to take this issue :)

**Comment by rcurtin** [comment]
Hi there, feel free!  If you submit a PR I'll look it over and we can go from there.

**Comment by MikeLing** [comment]
@rcurtin Hey, How could I just run discrete_distribution.cpp to test my pr and how could I use gdb or other debug tool about it? Thank you very much

**Comment by rcurtin** [comment]
If you mean that you want to run the discrete distribution tests, build the `mlpack_test` target and run the test suite:

```
$ bin/mlpack_test -t DiscreteDistributionTest
```

and the sources for that test are in `src/mlpack/tests/discrete_distribution_test.cpp`.

**Comment by MikeLing** [comment]
@rcurtin It said **Test setup error: no test cases matching filter or all test cases were disabled**

> MacBook-Pro:build mikeling$ bin/mlpack_test -t DiscreteDistributionTest
> Test setup error: no test cases matching filter or all test cases were disabled

How could I use gdb to test the special test like DiscreteDistributionTest? Because I want to get familiar with how exactly the discrete distribution works inside, so I want to debug the test case  step by step :)

**Comment by rcurtin** [comment]
Oh, sorry, I gave the wrong instructions.  The test suite is actually called `DistributionTest` not `DiscreteDistributionTest`, and the tests are found in `distribution_test.cpp`.  Take a look and you'll see what the names are.

So you could do something like, e.g.,

```
(gdb) break distribution_test.cpp:35
(gdb) run -t DistributionTest/DiscreteDistributionConstructorTest
```

You can see that I can also specify the name of the individual test to run, as well as the test suite.  So you can use this strategy to see how it is working.

**Comment by MikeLing** [comment]
Hey @rcurtin , I just create a PR[1] for this issue. I'm not sure if that's your need, but some code will be a good start I guess. And I will add comments if PR close to be merged :)

[1] https://github.com/mlpack/mlpack/pull/830

**Comment by Lakshya1605** [comment]
Is this still open ,if yes i would like to work on it .It may take some time  i am new here.Thank You

**Comment by MikeLing** [comment]
Hey @Lakshya1605, I'm working on it already. And I'm sure it's close to be done :)

**Comment by dineshraj01** [comment]
Hi, has this issue been resolved? I would like to work on it.

**Comment by MikeLing** [comment]
@dineshraj01 Sorry for the late reply, I meet some problems on hmm but I'm trying to resolve it. Could you wait a bit long? I will hand out this issue if I'm sure I can't finished it. Thank you :)

**Comment by rcurtin** [comment]
Fixed with #830.
